src/responder.ts

```diff
@@ -150,83 +176,329 @@ export class EthereumDedicatedResponder extends EthereumResponder {
+        );
+
+        // Get the initial gas price
+        this.gasPrice = await promiseTimeout(
+            this.signer.provider.getGasPrice(),
+            EthereumDedicatedResponder.WAIT_TIME_FOR_PROVIDER_RESPONSE
+        );
+
         while (this.mAttemptsDone < this.maxAttempts) {
             this.mAttemptsDone++;
             try {
                 // Try to call submitStateFunction, but timeout with an error if
-                // there is no response for 30 seconds.
-                const tx = await promiseTimeout(this.submitStateFunction(), 30000);
+                // there is no response for WAIT_TIME_FOR_PROVIDER_RESPONSE ms.
+                const tx = await promiseTimeout(
+                    this.submitStateFunction(responseData),
+                    EthereumDedicatedResponder.WAIT_TIME_FOR_PROVIDER_RESPONSE
+                );
 
                 // The response has been sent, but should not be considered confirmed yet.
-                this.responseFlow.status = ResponseState.ResponseSent;
-                this.asyncEmit(ResponderEvent.ResponseSent, this.responseFlow);
+                responseFlow.state = ResponseState.ResponseSent;
+                responseFlow.txHash = tx.hash;
+                this.asyncEmit(ResponderEvent.ResponseSent, responseFlow);
+
+
+                // Last block seen when transaction was first sent
+                const txSentBlockNumber = this.lastBlockNumberSeen;
+
+                // Promise that waits for the first confirmation
+                const firstConfirmationPromise = waitForConfirmations(this.signer.provider, tx.hash, 1);
+
+                // Promise that rejects after WAIT_BLOCKS_BEFORE_RETRYING blocks are mined since the transaction was first sent
+                const firstConfirmationTimeoutPromise = new Promise((_, reject) => {
+                    const testCondition = () => {
+                        if (this.lastBlockNumberSeen > txSentBlockNumber + EthereumDedicatedResponder.WAIT_BLOCKS_BEFORE_RETRYING) {
+                            reject(new StuckTransactionError(`Transaction still not mined after ${this.lastBlockNumberSeen - txSentBlockNumber} blocks`));
+                        }
+                        setTimeout(testCondition, 20);
+                    }
+                    testCondition();
+                });
+
+                // Promise that waits for enough confirmations before declaring success
+                const enoughConfirmationsPromise = waitForConfirmations(this.signer.provider, tx.hash, this.confirmationsRequired);
+
 
-                // Wait for enough confirmations before declaring success
-                const confirmationsPromise = waitForConfirmations(this.signer.provider, tx.hash, this.confirmationsRequired);
                 // ...but stop with error if no new blocks come for too long
+                // TODO: make sure this does not cause memory leaks
                 const noNewBlockPromise = new Promise((_, reject) => {
```

Comment:
This promise could be it's own class, allowing you to test it separately:

class NoNewBlocksPromise extends Promise<void> {

}

Reply:
Done with a function returning a CancellablePromise.